src/librustc_lint/builtin.rs

```diff
@@ -529,136 +529,96 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingDoc {
     }
 }
 
-declare_lint! {
-    pub MISSING_COPY_IMPLEMENTATIONS,
-    Allow,
-    "detects potentially-forgotten implementations of `Copy`"
-}
-
-#[derive(Copy, Clone)]
-pub struct MissingCopyImplementations;
-
-impl LintPass for MissingCopyImplementations {
-    fn name(&self) -> &'static str {
-        "MissingCopyImplementations"
-    }
-
-    fn get_lints(&self) -> LintArray {
-        lint_array!(MISSING_COPY_IMPLEMENTATIONS)
-    }
-}
+macro_rules! missing_impl {
+    ($trait:ident, $trait_path:path, $lint_constant:ident, $struct_name:ident,
+        $trait_check:ident) => (
+        declare_lint! {
+            pub $lint_constant,
+            Allow,
+            concat!("detects missing implementations of ", stringify!($trait_path))
+        }
 
-impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingCopyImplementations {
-    fn check_item(&mut self, cx: &LateContext, item: &hir::Item) {
-        if !cx.access_levels.is_reachable(item.id) {
-            return;
+        pub struct $struct_name {
+            impling_types: Option<NodeSet>,
         }
-        let (def, ty) = match item.node {
-            hir::ItemKind::Struct(_, ref ast_generics) => {
-                if !ast_generics.params.is_empty() {
-                    return;
-                }
-                let def = cx.tcx.adt_def(cx.tcx.hir().local_def_id(item.id));
-                (def, cx.tcx.mk_adt(def, cx.tcx.intern_substs(&[])))
-            }
-            hir::ItemKind::Union(_, ref ast_generics) => {
-                if !ast_generics.params.is_empty() {
-                    return;
-                }
-                let def = cx.tcx.adt_def(cx.tcx.hir().local_def_id(item.id));
-                (def, cx.tcx.mk_adt(def, cx.tcx.intern_substs(&[])))
-            }
-            hir::ItemKind::Enum(_, ref ast_generics) => {
-                if !ast_generics.params.is_empty() {
-                    return;
-                }
-                let def = cx.tcx.adt_def(cx.tcx.hir().local_def_id(item.id));
-                (def, cx.tcx.mk_adt(def, cx.tcx.intern_substs(&[])))
+
+        impl $struct_name {
+            pub fn new() -> $struct_name {
+                $struct_name { impling_types: None }
             }
-            _ => return,
-        };
-        if def.has_dtor(cx.tcx) {
-            return;
         }
-        let param_env = ty::ParamEnv::empty();
-        if ty.is_copy_modulo_regions(cx.tcx, param_env, item.span) {
-            return;
-        }
-        if param_env.can_type_implement_copy(cx.tcx, ty).is_ok() {
```

Comment:
I don't think this warned you if the type couldn't be `Copy`, it just didn't trigger the lint. I'm wondering if this is really a problem in practice. I'd think that if you set this lint globally, you'd want it to trigger on all types and not exclude types that it can't be implemented for, so you'd have to whitelist them explicitly.

Reply:
I think this is super important.  I would never use a lint suggesting `Copy` if I had to suppress it any time I put a `Vec` or a `HashSet` inside my struct.